tests/test_pytorch/test_fsdp.py

```diff
@@ -0,0 +1,791 @@
+
+    def configure_model(self) -> None:
+        if self.layer is None:
+            self._init_model()
+        # the model is already wrapped with FSDP: no need to wrap again!
+        if isinstance(self.layer, FullyShardedDataParallel):
+            return
+
+        self.layer = wrap(self.layer)
+
+    def configure_optimizers(self):
+        # There is some issue with SGD optimizer state in FSDP
+        return torch.optim.AdamW(self.layer.parameters(), lr=0.1)
+
+    def on_train_batch_start(self, batch, batch_idx):
+        assert batch.dtype == torch.float32
+
+    def on_train_batch_end(self, _, batch, batch_idx):
+        assert batch.dtype == torch.float32
+        self._assert_layer_fsdp_instance()
+
+    def on_test_batch_end(self, _, batch, batch_idx):
+        assert batch.dtype == torch.float32
+        self._assert_layer_fsdp_instance()
+
+    def on_validation_batch_end(self, _, batch, batch_idx):
+        assert batch.dtype == torch.float32
+        self._assert_layer_fsdp_instance()
+
+    def on_predict_batch_end(self, _, batch, batch_idx):
+        assert batch.dtype == torch.float32
+        self._assert_layer_fsdp_instance()
+
+    def _assert_layer_fsdp_instance(self) -> None:
+        assert isinstance(self.layer, FullyShardedDataParallel)
+        assert isinstance(self.trainer.strategy.precision_plugin, HPUFSDPPrecision)
+
+        if self.trainer.precision == "16-mixed":
+            param_dtype = None if not _TORCH_GREATER_EQUAL_2_0 else torch.float32
+            reduce_dtype = buffer_dtype = torch.float16
+        elif self.trainer.precision == "bf16-mixed":
+            param_dtype = None if not _TORCH_GREATER_EQUAL_2_0 else torch.float32
+            reduce_dtype = buffer_dtype = torch.bfloat16
+        elif self.trainer.precision == "16-true":
+            param_dtype = reduce_dtype = buffer_dtype = torch.float16
+        elif self.trainer.precision == "bf16-true":
+            param_dtype = reduce_dtype = buffer_dtype = torch.bfloat16
+        else:
+            raise ValueError(f"Unknown precision {self.trainer.precision}")
+
+        assert self.layer.mixed_precision.param_dtype == param_dtype
+        assert self.layer.mixed_precision.reduce_dtype == reduce_dtype
+        assert self.layer.mixed_precision.buffer_dtype == buffer_dtype
+
+
+class TestBoringModel(BoringModel):
+    def __init__(self, wrap_min_params: int = 2):
+        super().__init__()
+
+        self.save_hyperparameters()
+        self.layer = torch.nn.Sequential(torch.nn.Linear(32, 32), torch.nn.ReLU(), torch.nn.Linear(32, 2))
+        self.should_be_wrapped = [wrap_min_params < (32 * 32 + 32), None, wrap_min_params < (32 * 2 + 2)]
+
+    def configure_optimizers(self):
+        parameters = self.parameters() if _TORCH_GREATER_EQUAL_2_0 else self.trainer.model.parameters()
+
+        # SGD's FSDP optimizer state is fixed in https://github.com/pytorch/pytorch/pull/99214
```

Comment:
you can add this to separte sgd tests if required

Reply:
Removed. This test is taken as is from upstream lightning. I think the issue is resolved now https://github.com/pytorch/pytorch/pull/99214. I will add another test with SGD in a separate PR.